src/input.rs

```diff
@@ -1,98 +1,41 @@
-            self.replacer.replace_preview(&buffer)
-        } else {
-            self.replacer.replace(&buffer)
-        })?;
-
-        Ok(())
+impl Source {
+    pub(crate) fn from_paths(paths: Vec<PathBuf>) -> Vec<Self> {
+        paths.into_iter().map(Self::File).collect()
     }
 
-    pub(crate) fn new(source: Source, replacer: Replacer) -> Self {
-        Self { source, replacer }
+    pub(crate) fn from_stdin() -> Vec<Self> {
+        vec![Self::Stdin]
     }
-    pub(crate) fn run(&self, preview: bool) -> Result<()> {
-        let is_tty = std::io::stdout().is_terminal();
-
-        match (&self.source, preview) {
-            (Source::Stdin, true) => self.stdin_replace(is_tty),
-            (Source::Stdin, false) => self.stdin_replace(is_tty),
-            (Source::Files(paths), false) => {
-                use rayon::prelude::*;
-
-                let failed_jobs: Vec<_> = paths
-                    .par_iter()
-                    .filter_map(|p| {
-                        if let Err(e) = self.replacer.replace_file(p) {
-                            Some((p.to_owned(), e))
-                        } else {
-                            None
-                        }
-                    })
-                    .collect();
 
-                if failed_jobs.is_empty() {
-                    Ok(())
-                } else {
-                    let failed_jobs =
-                        crate::error::FailedJobs::from(failed_jobs);
-                    Err(Error::FailedProcessing(failed_jobs))
-                }
-            }
-            (Source::Files(paths), true) => {
-                let stdout = std::io::stdout();
-                let mut handle = stdout.lock();
-                let print_path = paths.len() > 1;
-
-                paths.iter().try_for_each(|path| {
-                    if Replacer::check_not_empty(File::open(path)?).is_err() {
-                        return Ok(());
-                    }
-                    let file =
-                        unsafe { memmap2::Mmap::map(&File::open(path)?)? };
-                    if self.replacer.has_matches(&file) {
-                        if print_path {
-                            writeln!(
-                                handle,
-                                "----- FILE {} -----",
-                                path.display()
-                            )?;
-                        }
-
-                        handle
-                            .write_all(&self.replacer.replace_preview(&file))?;
-                        writeln!(handle)?;
-                    }
-
-                    Ok(())
-                })
-            }
+    pub(crate) fn display(&self) -> String {
+        match self {
+            Self::Stdin => "STDIN".to_string(),
+            Self::File(path) => format!("FILE {}", path.display()),
         }
     }
 }
+
+pub(crate) fn make_mmap(path: &PathBuf) -> Result<Mmap> {
+    Ok(unsafe { Mmap::map(&File::open(path)?)? })
+}
```

Comment:
```suggestion
pub(crate) unsafe fn make_mmap(path: &PathBuf) -> Result<Mmap> {
    Ok(unsafe { Mmap::map(&File::open(path)?)? })
}
```

I think crate internal unsoundness is "fine", but `make_mmap()` has invariants that need to be upheld from the internal `Mmap::map()`. It should really be `unsafe` too

Reply:
Yes, I agree that there are more precautions we should take when `Mmap`ing the file like locking the file, but I'd prefer if we could handle that in a separate PR. I just want to retain that this function call is unsafe if we're moving mmap creation to it's own function

Reply:
Alright, I'll add a TODO and push it.